Response.json method change

yisraelU opened the pull request:
First of all , awesome library . thank you
changed the method to take in CharSequence instead of String.
provides better performance,flexibility plus compatibility with zio-json which encodes to charsequence
Please let me know if this implementation is acceptable , any feedback is desired

yisraelU committed:
changed Response.json to take in any CharSequence

yisraelU committed:
Merge branch 'dream11:main' into main

jdegoes commented on zio-http/src/main/scala/zhttp/http/Response.scala:
```diff
@@ -180,9 +180,9 @@ object Response {
+  def json(data: CharSequence): Response =
     Response(
-      data = HttpData.fromChunk(Chunk.fromArray(data.getBytes(HTTP_CHARSET))),
+      data = HttpData.fromString(data.toString),
```
Would be good to push this change into `Http.fromString`, all the way down as low as possible.

tusharmath replied:
We can add a new primitive for `HttpData.fromCharSequence` and `Http.fromString` can also use it since `String <: CharSequence`

yisraelU replied:
@jdegoes , @tusharmath is there a need for 2 primitives? perhaps just change HttpData.Text to hold a CharSequence

jdegoes replied:
I think that's correct, since `CharSequence` is always more efficient than `String`.

yisraelU replied:
I pushed change for review . I can add a new method fromCharSequence to replace the fromString method and mark fromString deprecated as it will be subsumed under the method fromCharSequence.

jdegoes replied:
I guess the main problem is people will look for `fromString`, expecting it to be there. Not sure.

yisraelU committed:
Merge branch 'dream11:main' into main

yisraelU committed:
changed HttpData.Text to hold CharSequence.

yisraelU committed:
Merge remote-tracking branch 'origin/main'

yisraelU committed:
fixed mistaken push

yisraelU committed:
Merge branch 'dream11:main' into main

tusharmath commented on zio-http/src/main/scala/zhttp/http/HttpData.scala:
```diff
@@ -80,14 +80,14 @@ object HttpData {
   def fromFile(file: java.io.File): HttpData = File(file)
 
-  private[zhttp] final case class Text(text: String, charset: Charset)                   extends HttpData
+  private[zhttp] final case class Text(text: CharSequence, charset: Charset)             extends HttpData
```
This is good.

tusharmath commented on zio-http/src/main/scala/zhttp/http/HttpData.scala:
```diff
@@ -80,14 +80,14 @@ object HttpData {
    * Helper to create HttpData from String
    */
-  def fromString(text: String, charset: Charset = HTTP_CHARSET): HttpData = Text(text, charset)
+  def fromString(text: CharSequence, charset: Charset = HTTP_CHARSET): HttpData = Text(text, charset)
```
Keep this, and let's add a `fromCharSequence` also.

yisraelU replied:
Ok great

tusharmath commented on zio-http/src/main/scala/zhttp/http/Response.scala:
```diff
@@ -86,7 +81,7 @@ final case class Response private (
     val jHeaders = self.getHeaders.encode
     val jContent = self.data match {
-      case HttpData.Text(text, charset) => Unpooled.wrappedBuffer(text.getBytes(charset))
+      case HttpData.Text(text, charset) => Unpooled.copiedBuffer(text, charset)
```
Is this change required?

yisraelU replied:
@tusharmath 
wrappedbuffer doesn't take in CharSequence .
we can do text.toString.getBytes instead .
Is that preferred?
String.getBytes performs a copy either way

tusharmath replied:
Got it. May be we can do something like this.

```scala
case HttpData.Text(ascii: AsciiString, _) => Unpooled.wrappedBuffer(ascii)
case HttpData.Text(text, charset)         => Unpooled.copiedBuffer(text, charset)
```

The overall idea is to reduce the copies we create as much as possible. This makes a big difference in performance when we send large payloads to the clients.

yisraelU replied:
ok , ill look into it

tusharmath commented on zio-http/src/main/scala/zhttp/http/Response.scala:
```diff
@@ -185,9 +180,9 @@ object Response {
    * Creates a response with content-type set to application/json
    */
-  def json(data: String): Response =
+  def json(data: CharSequence): Response =
```
We can do this for `Response.text("abc")` also.

yisraelU replied:
I'll add

tusharmath commented on zio-http/src/main/scala/zhttp/service/Client.scala:
```diff
@@ -156,7 +156,7 @@ object Client {
   ) extends HeaderExtension[ClientRequest] { self =>
 
     def getBodyAsString: Option[String] = data match {
```
We can add `getBodyAsCharSequence` also. It's alright if you want to do it in a different PR :)

yisraelU replied:
done

yisraelU committed:
new method HttpData.fromCharSequence
Changed Response.text to take in CharSequence
Response.json/text use new HttpData.fromCharSequence method
new method getBodyAsCharSequence

tusharmath commented:
Can you take a pull from main.

yisraelU committed:
merge

yisraelU committed:
merged

yisraelU committed:
merge

yisraelU committed:
added AsciiString case

yisraelU committed:
Merge branch 'main' of https://github.com/dream11/zio-http into dream11-main

# Conflicts:
#	zio-http/src/main/scala/zhttp/service/Client.scala

yisraelU committed:
Merge branch 'dream11-main'

yisraelU commented:
> Can you take a pull from main.

@tusharmath done.

tusharmath commented on zio-http/src/main/scala/zhttp/service/Client.scala:
```diff
@@ -175,6 +175,22 @@ object Client {
     self =>
 
+    def bodyAsString: Task[String] = getBodyAsByteBuf.map(_.toString(headers.charset))
+    def getBodyAsString: Option[String] = data match {
```
```suggestion
    def getBodyAsString: Task[String] = data match {
```

tusharmath commented on zio-http/src/main/scala/zhttp/service/Client.scala:
```diff
@@ -175,6 +175,22 @@ object Client {
+      case _                            => Option.empty
+    }
+
+    def getBodyAsCharSequence: Option[CharSequence] = data match {
```
```suggestion
    def getBodyAsCharSequence: Task[CharSequence] = data match {
```

tusharmath commented on zio-http/src/main/scala/zhttp/http/HttpData.scala:
```diff
@@ -77,6 +77,11 @@ object HttpData {
+  /**
+   * Helper to create HttpData from a CharSequence
+   */
+  def fromCharSequence(text: CharSequence, charset: Charset = HTTP_CHARSET): HttpData = Text(text, charset)
```
```suggestion
  def fromCharSequence(text: CharSequence): HttpData = Text(text, charset)
```

I think `fromCharSequence` doesn't need `Charset`

tusharmath commented on zio-http/src/main/scala/zhttp/http/HttpData.scala:
```diff
@@ -87,7 +92,7 @@ object HttpData {
   def fromFile(file: java.io.File): HttpData = File(file)
 
-  private[zhttp] final case class Text(text: String, charset: Charset)                   extends HttpData
+  private[zhttp] final case class Text(text: CharSequence, charset: Charset)             extends HttpData
```
```suggestion
  private[zhttp] final case class Text(text: CharSequence)             extends HttpData
```

tusharmath commented:
Expecting many more conflicts from main. So closing this in favour of #1142
